add new documentation guidelines (#14115)
paritytech/substrate: docs/DOCUMENTATION_GUIDELINES.md

//!
//! <Your pallet must have a few tests that cover important user journeys. Use https://crates.io/crates/docify to reuse your these as examples>.
//!
//! ## Code Details
//!
//! <inside [`pallet`] module, a template is auto-generated that leads the reader to the relevant items. This is just a reminder to navigate to that page. Don't repeat things like "See Config trait for ..." that are generated inside [`pallet`] in here>
//!
//! See [`pallet`] module.
//!
//! <The audience of this is those who want to know how this pallet works, to the extent of being able to build something on top of it, like a DApp or another pallet>
//!
//! ## Low Level / Implementation Details
//!
//! <The format of this section is up to you, but we suggest the Design-oriented approach that follows>
//!
//! <The audience of this would be your future self, and those who want to deeply understand the pallet to optimize it further and such>
//!
//! ### Design Goals (optional)
//!
//! <Describe your goals with with the pallet design.>
//!
//! ### Design (optional)
//!
//! <Describe how you reach those goals. This should namely describe how your storage is laid out.>
//!
//! ### Terminology (optional)
//!

> //! <Optionally, contain any custom terminology here. You can link to it if you want to use the terminology further up>

Reviewer: ```suggestion
//! <Optionally, explain any custom terminology here. You can link to it if you want to use the terminology further up>
```

### TLDR

0. Have the goal of enforcing `#![deny(missing_docs)]` mentally, even if it is not enforced by the compiler 🙈.
1. Start with a single, clear and concise sentence. Follow up with more context, after a newline, if needed.
2. Use examples as much as reasonably possible.
3. Use links as much as possible.
4. Think about context. If you are in need of explaining a lot of foreign topics in documenting a trait that should not explicitly depend on them, you have likely not designed it properly.

### Other Guidelines

The above five guidelines must always be reasonably respected in the documentation.

The following are a set of notes that may not necessarily hold in all circumstances:

#### Document Through Code

You should make sure that your code is properly-named and well-organized so that your code functions as a form of documentation. However, within the complexity of our projects in Polkadot/Substrate that is not enough. Particularly, things like examples, errors and panics cannot be documented only through properly-named and well-organized code.

> Our north star is self-documenting code that also happens to be well-documented and littered with examples.

> * This is hard to scale at a project the size of Polkadot/Substrate. Moreover, things like examples, errors and panics cannot be documented via just code.

Author: ```suggestion
```

# Substrate Documentation Guidelines

> This document is only focused on documenting parts of substrate that relates to its external API. The list of such crates can be found in [CODEOWNERS](./CODEOWNERS). Search for the crates that are auto-assigned to a team called `docs-audit`.

Author: I would defer this decision to be made as a part of https://github.com/paritytech/labels/pull/29 by @the-right-joyce and @juangirini. As far as I know, the agreed that using labels for things like audit was not the right choice as it is not a permanent label (eg 'needs audit' -> 'audited').

> This document is only focused on documenting parts of substrate that relates to its external API. The list of such crates can be found in [CODEOWNERS](./CODEOWNERS). Search for the crates that are auto-assigned to a team called `docs-audit`.

Author: I find the label be a redundant side effect of the auto-review rule, isn't that the case? we already will assign this team to review automatically. I think this is semantically the same as the label.

> This document is only focused on documenting parts of substrate that relates to its external API. The list of such crates can be found in [CODEOWNERS](./CODEOWNERS). Search for the crates that are auto-assigned to a team called `docs-audit`.

Reviewer 2: True, good point. Though the auto-review rule will only ask the @paritytech/docs-audit team for a review. Should we add a rule to require at least one approval from them on top of the approvals we already need?

> This document is only focused on documenting parts of substrate that relates to its external API. The list of such crates can be found in [CODEOWNERS](./CODEOWNERS). Search for the crates that are auto-assigned to a team called `docs-audit`.

Author: Eventually, yes, but let's first see if the docs-audit team has enough bandwidth for this.

> This document is only focused on documenting parts of substrate that relates to its external API. The list of such crates can be found in [CODEOWNERS](./CODEOWNERS). Search for the crates that are auto-assigned to a team called `docs-audit`.

Reviewer 3: we got rid off the audit labels, for the security audits we will use Github projects in future. The process still needs to be documented, but as a preparation I already have here the project set up: https://github.com/orgs/paritytech/projects/103
I'm syncing with Regina from SRLabs regarding that.

When it comes to the docs update, I'd also say this should be part of the review process. The same way we're changing the release process and want reviewers to add/adapt the changelog if needed, I think it should also be in their responsibility to think if the PR is important for docs as well and respectively add @paritytech/docs-audit as reviewers and the docs team are the ones who decide if the docs should be updated or not - does that make sense?

> These are crates that are often used by external developers and deserve a better documents. Consequently, most of this revolves around FRAME-development.

Reviewer 4: ```suggestion
These are crates that are often used by external developers and need more thorough documentation. These are the crates most concerned with FRAME development.
```

- [Substrate Documentation Guidelines](#substrate-documentation-guidelines)
  - [General/Non-Pallet Crates](#generalnon-pallet-crates)
    - [What to Document?](#what-to-document)
      - [Rust Docs vs. Code Comments](#rust-docs-vs-code-comments)
    - [How to Document?](#how-to-document)
    - [TLDR](#tldr)
    - [Other Guidelines](#other-guidelines)
      - [Document Through Code](#document-through-code)
      - [Formatting Matters](#formatting-matters)
  - [Pallet Crates](#pallet-crates)
    - [Top Level Pallet Docs (`lib.rs`)](#top-level-pallet-docs-librs)
    - [Dispatchable](#dispatchable)
    - [Storage Items](#storage-items)

## General/Non-Pallet Crates

First, consider the case for all such crates, except for those that are pallets.

### What to Document?

The first question is, what should you document? Use the following filter:

> 1. Within the set of crates above,

Reviewer 4: Remove this line, it's unnecessary.

Author: I kinda disagree 🙈 I think it is good to remind people that this only applies to certain crates. Instead, I will make it a bit cleared. Let me know what you think.

> 2. Any `pub` item within them needs to be documented. If it is not `pub`, it won't show up in the rust-docs, and won't matter.

Reviewer 4: ```suggestion
2. All `pub` item within the aforementioned crates need to be documented. If it is not `pub`, it does not appear in the rust-docs, and is not public facing.
```

Reviewer: I suggest adding a final sentence: "Documentation should be encouraged, but not be mandatory for non-trivial private items to improve code readability.

>     * Within `pub` items, sometimes they are only `pub` in order to be used by another internal crate, and you can foresee that this will not be used by anyone else other than you. These need **not** be documented thoroughly, and are left to your discretion to identify.

Reviewer 4: ```suggestion
    * There are some `pub` items that are only `pub` in order to be accessible for another internal crate. These should still be documented although you do not need to be as thorough as you are with the `pub` items that are for external developer use.
```

Reviewer 5: Should we ask to add a comment like "//no-doc" to establish that the author intentionally didn't add the docs?

Author: @hummusonrails I am a bit hesitant about your suggestions that seem paraphrasing to me. Do you have strong opinions that the original text is wrong or poorly worded, or are you adding optional suggestions? I accepted most, but might close some. Please let me know if you prefer otherwise.

Reviewer 4: @kianenigma In this one, I attempted to make it more neutral and formal, whereas your version was more informal and personal in tone. I'm of the opinion that documentation should lean more into the former. That's why I made that suggestion. Hope that helps clarify!

    * Reminder: `trait` items are public by definition, if the trait is public.

> 3. Of course, all of these crates should have a reasonable module documentation (`//!`).

Reviewer 4: ```suggestion
3. All crates should have reasonable module documentation (`//!`).
```

Author: edited your suggestion to actually be more correct, neither of us are right 🙈  crates don't have `mod` docs, modules do. And all public modules should have docs.

#### Rust Docs vs. Code Comments

> Note that anything starting with `///` is external rust-doc, and anything starting with `//` is just for you to see. Don't mix the two!

Reviewer 4: ```suggestion
Note that anything starting with `///` is an external rust-doc, and everything starting with `//` does not appear in the rust-docs. It's important to not confuse the two in your documentation.
```

```rust
/// Computes the square root of the input, returning `Ok(_)` if successful.
///
/// # Errors
/// ...
///

> /// # Complexity

Author: ```suggestion
/// # Return Type
```

/// ...
///

> // Details about the complexity, how you implemented this, and some quirks that

Reviewer 5: We created the section `#Complexity` above but are saying that the details about complexity should start with `//`.

Reviewer: I think Kian means to note the complexity in the `///` comment, but leave the gritty details of the complexity (why, trade-offs between different considered approaches, etc) to the `//` comment.

Author: I meant to say more details of how the implementation is achieved and so on, but I get that this is very confusing 🙈 edited.

> /// # Complexity
> /// ...
> ///

Author: ```suggestion
```

// are NOT relevant to the external interface, so it starts with '//'.
// This can also be moved inside the function.
pub fn sqrt(x: u32) -> Result<u32, ()> {
    todo!();
}

```

### How to Document?

There are a few very good sources that you can look into:

- https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html
- https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/documentation.html
- https://blog.guillaume-gomez.fr/articles/2020-03-12+Guide+on+how+to+write+documentation+for+a+Rust+crate

> As mentioned [here](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/documentation.html#writing-documentation-comments) and [here](https://blog.guillaume-gomez.fr/articles/2020-03-12+Guide+on+how+to+write+documentation+for+a+Rust+crate), always start with a **single sentence** demonstrating what is being documented. If there is more to say, add it *after a newline*.

Reviewer 4: ```suggestion
As mentioned [here](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/documentation.html#writing-documentation-comments) and [here](https://blog.guillaume-gomez.fr/articles/2020-03-12+Guide+on+how+to+write+documentation+for+a+Rust+crate), always start with a **single sentence** demonstrating what is being documented. All additional documentation should be added *after a newline*. Strive to make the first sentence succinct and short.
```

> About [special sections](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/documentation.html#special-sections), we will most likely not need to think about panic and safety in any runtime related code. Our code is never `unsafe`, and will (almost) never panic.

Reviewer 4: ```suggestion
About [special sections](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/documentation.html#special-sections), we will most likely not need to think about panic and safety in any runtime related code. Our code is never `unsafe`, and will (almost) never panic. If you see a panic in the code that is a reason to raise a question to the developer who introduced it rather than work to document it.
```

Author: can you elaborate more what you are suggesting? in general, you might see things like `expect()` in the code but that is acceptable. Did you take that into account?

> Use `# Example`s as much as possible. These are great ways to further demonstrate what your APIs are doing, and add free test coverage. As a cherry on top, any code in rust-docs is treated as an "integration tests", not unit tests, which tests your crate in a different way than unit tests. So, it is both a win for "more documentation" and a win for "more test coverage".

Reviewer 4: ```suggestion
Use `# Examples as much as possible. These are great ways to further demonstrate what your APIs are doing, and add free test coverage. As an additional benefit, any code in rust-docs is treated as an "integration tests", not unit tests, which tests your crate in a different way than unit tests. So, it is both a win for "more documentation" and a win for "more test coverage".
```

> Use `# Example`s as much as possible. These are great ways to further demonstrate what your APIs are doing, and add free test coverage. As a cherry on top, any code in rust-docs is treated as an "integration tests", not unit tests, which tests your crate in a different way than unit tests. So, it is both a win for "more documentation" and a win for "more test coverage".

Reviewer 4: As a general good practice avoid using any slang or expressions in docs as non-native speakers or speakers from different regions may not know that term and it introduces unnecessary complexity

Author: noted, thanks!

> You can also consider having am `# Error` section, but consider that optional. Of course, this only applies if there is a `Result` being returned, and IFF the `Error` variants are overly complicated.

Reviewer 6: ```suggestion
You can also consider having an `# Error` section, but consider that optional. Of course, this only applies if there is a `Result` being returned, and IFF the `Error` variants are overly complicated.
```

> You can also consider having am `# Error` section, but consider that optional. Of course, this only applies if there is a `Result` being returned, and IFF the `Error` variants are overly complicated.

Reviewer 4: ```suggestion
You can also consider having an `# Error` section optionally. Of course, this only applies if there is a `Result` being returned, and if the `Error` variants are overly complicated.
```

> Strive to include correct links to other items in your written docs as much as possible. In other words, avoid \`some_func\` and instead use \[\`some_fund\`\]. Read more about how to correctly use links in your rust-docs [here](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html#valid-links) and [here](https://rust-lang.github.io/rfcs/1946-intra-rustdoc-links.html#additions-to-the-documentation-syntax).

Reviewer 5: \[\`some_func\`\]

Author: I think Sam corrected me properly below 🙈

> Strive to include correct links to other items in your written docs as much as possible. In other words, avoid \`some_func\` and instead use \[\`some_fund\`\]. Read more about how to correctly use links in your rust-docs [here](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html#valid-links) and [here](https://rust-lang.github.io/rfcs/1946-intra-rustdoc-links.html#additions-to-the-documentation-syntax).

Reviewer 7: ```suggestion
Strive to include correct links to other items in your written docs as much as possible. In other words, avoid `` `some_func` `` and instead use ``[`some_func`]``. Also note that there are special considerations when documenting proc macros. Doc links will appear to function _within_ your proc macro crate, but often will no longer function when these proc macros are re-exported elsewhere in your project. The exception is doc links to _other proc macros_ which will function just fine if they are also being re-exported. It is also often necessary to disambiguate between a proc macro and a function of the same name, which can be done using the `macro@my_macro_name` syntax in your link. Read more about how to correctly use links in your rust-docs [here](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html#valid-links) and [here](https://rust-lang.github.io/rfcs/1946-intra-rustdoc-links.html#additions-to-the-documentation-syntax).
```

> > While you are linking, you might become conscious of the fact that you are in need of linking to (too many) foreign items in order to explain your API. This is leaning more towards API-Design rather than documentation, but it is a warning that the subject API might be slightly wrong. For example, most "glue" traits[^1] in `frame/support` should be designed, and documented without making hard assumptions about a particular pallets that implement them.

Reviewer 4: ```suggestion
> While you are linking, you might become conscious of the fact that you are in need of linking to (too many) foreign items in order to explain your API. This is leaning more towards API-Design rather than documentation, but it is a warning that the subject API might be slightly wrong. For example, most "glue" traits[^1] in `frame/support` should be designed and documented without making hard assumptions about particular pallets that implement them.
```

### TLDR

0. Have the goal of enforcing `#![deny(missing_docs)]` mentally, even if it is not enforced by the compiler 🙈.
1. Start with a single, clear and concise sentence. Follow up with more context, after a newline, if needed.
2. Use examples as much as reasonably possible.
3. Use links as much as possible.

> 4. Think about context. If you are in need of explaining a lot of foreign topics in documenting a trait that should not explicitly depend on them, you have likely not designed it properly.

Reviewer: ```suggestion
4. Think about context. If you are explaining a lot of foreign topics while documenting a trait that should not explicitly depend on them, you have likely not designed it properly.
```

### Other Guidelines

> The above 5 guidelines must always be respected in docs, as long as reasonable possible.

Reviewer 4: Too much equivocation here in my opinion.

```suggestion
The above five guidelines must always be reasonably respected in the documentation.
```

> The following are a set of notes that may not necessarily hold at all circumstances. Consider them as opinionated guidelines that you may or may not abide by.

Reviewer 4: ```suggestion
The following are a set of notes that may not necessarily hold at all circumstances:
```

> The following are a set of notes that may not necessarily hold at all circumstances. Consider them as opinionated guidelines that you may or may not abide by.

Reviewer 7: ```suggestion
The following are a set of notes that may not necessarily hold in all circumstances:
```

"at" is the wrong preposition here, should be "in"

#### Document Through Code

> You can certainly do parts of your documentation through your well-organized, properly-named code. That's all true, but:

Reviewer 4: ```suggestion
You should make sure that your code is properly-named and well-organized so that your code functions as a form of documentation. However, within the complexity of our projects in Polkadot/Substrate that is not enough. Particularly, things like examples, errors and panics cannot be documented only through properly-named and well-organized code.
```

Reviewer 7: I would even add after that:

```
Our north star is self-documenting code that also happens to be well-documented and littered with examples.
```

> * This is hard to scale at a project the size of Polkadot/Substrate. Moreover, things like examples, errors and panics cannot be documented via just code.

Reviewer 4: Remove this sentence.

> * Your written documents should *complement* the code, not *repeat* it. Put bluntly, if you end up writing this, you are likely doing it wrong:

Reviewer 4: ```suggestion
* Your written documents should *complement* the code, not *repeat* it. As an example, a documentation on top of a code example should never look like the following:
```

 ```rust
 /// Sends request and handles the response.
 trait SendRequestAndHandleResponse {

 }
 ```

>  Because the document is adding no extra information and you are better of without it.

Reviewer 4: ```suggestion
In the above example, the documentation has added no useful information not already contained within the properly-named trait and is redundant. 
```

#### Formatting Matters

> The way you format your documents (newlines, heading and so on) do matter! Consider the below examples:

Reviewer 4: ```suggestion
The way you format your documents (newlines, heading and so on) make a difference. Consider the below examples:
```

Reviewer 2: It should be 'makes a difference'

Reviewer: or 'makes a difference'?

Reviewer: or 'makes a difference'?

```rust

> /// This function works with input u32 x and multiples it by two. If

Reviewer 2: ```suggestion
/// This function works with input u32 x and multiplies it by two. If
```

/// we optimize the other variant of it, we would be able to achieve more
/// efficiency but I have to think about it. Probably can panic if the input
/// overflows u32.
fn multiply_by_2(x: u32) -> u32 { .. }
```

```rust
/// Multiplies an input of type [`u32`] by two.
///
/// # Panics
///
/// Panics if the input overflows.
///
/// # Complexity
///
/// Is implemented using some algorithm that yields complexity of O(1).
// More efficiency can be achieved if we improve this via such and such.
fn multiply_by_2(x: u32) -> u32 { .. }
```

> They are both roughly conveying the same set of facts, but one is substantially kinder on the eye. Especially for traits and type that you can foresee will be seen/used a lot, try and write the better version!

Reviewer 4: ```suggestion
They are both roughly conveying the same set of facts, but one is easier to follow because it was formatted cleanly. Especially for traits and type that you can foresee will be seen and used a lot, try and write a well formatted version.
```

> Similarly, make sure your comments are wrapped at 100 characters line-width (as defined by our [`rustfmt.toml`](../rustfmt.toml)), no **more and no less**! The more is fixed by `rustfmt` and our CI, but if you (some some unknown reason) wrap your lines at 59 characters, it will pass the CI, and it will not look good 🫣.

Reviewer 7: ```suggestion
Similarly, make sure your comments are wrapped at 100 characters line-width (as defined by our [`rustfmt.toml`](../rustfmt.toml)), no **more and no less**! The more is fixed by `rustfmt` and our CI, but if you (some some unknown reason) wrap your lines at 59 characters, it will pass the CI, and it will not look good 🫣. Note that the [rewrap](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) Visual Studio Code extension makes it quite easy to re-wrap any piece of code/comments/documentation at a configurable column boundary, and is used by many of our core devs especially when working on documentation.
```

Reviewer: A benefit I've noticed `rewrap` has over `rustfmt` is it'll collapse two lines up into one line if they can both fit, as well as wrap one line onto another if it exceeds the max line len.

Maybe I'm missing an option with `rustfmt` though? It would be cool if we could achieve this without the extra tool.

Reviewer 7: rustfmt will _sort of_ do it most of the time, but it has fewer types of lines it knows how to word wrap, especially when it comes to markdown within rust doc comments. As @liamaharon said it will also condense things when it detects it can, so when you have a paragraph of doc comment, and then you insert or remove a word from the first sentence, it will handle the cascading word wrap change and optimally re-wrap the entire paragraph if you select it and ask it to.

Shawn was a big fan of it

Author: rustfmt, as noted by Liam, does not fix this, and it drives me crazy sometimes 🙈 

https://github.com/paritytech/ci_cd/issues/626

> Similarly, make sure your comments are wrapped at 100 characters line-width (as defined by our [`rustfmt.toml`](../rustfmt.toml)), no **more and no less**! The more is fixed by `rustfmt` and our CI, but if you (some some unknown reason) wrap your lines at 59 characters, it will pass the CI, and it will not look good 🫣.

Reviewer: ```suggestion
Similarly, make sure your comments are wrapped at 100 characters line-width (as defined by our [`rustfmt.toml`](../rustfmt.toml)), no **more and no less**! The more is fixed by `rustfmt` and our CI, but if you (for some unknown reason) wrap your lines at 59 characters, it will pass the CI, and it will not look good 🫣.
```

[^1]: Those that help two pallets talk to each other.

## Pallet Crates

> Everything above is related to non-pallet details. They might be relevant in both crates that are pallets, and non-pallet crates.

Reviewer: ```suggestion
The guidelines so far have been general in nature, and are applicable to crates that are pallets and crates that're not pallets.
```

The following is relevant to how to document parts of a crate that is a pallet.

### Top Level Pallet Docs (`lib.rs`)

> for the top-level pallet docs, consider the following template:

Reviewer 6: Can you link to the fast-unstake as an example?

Author: Done ✅

> for the top-level pallet docs, consider the following template:

Reviewer 2: ```suggestion
For the top-level pallet docs, consider the following template:
```

```

> //! # <Pallet Name>

Reviewer 4: This whole template is really nicely done!

//!
//! <single-liner about the pallet>.
//!
//! ## High-Level/End-User Details
//!
//! <a few paragraphs, focus on what external folks should know about the pallet>
//!
//! <The audience here is potentially non-coders who just want to know what this pallet does, not how it does it>
//!
//! ### Example
//!
//! <Your pallet must have a few tests that cover important user journeys. Use https://crates.io/crates/docify to reuse your these as examples>.
//!
//! ## Code Details
//!

> //! <inside [`pallet`] module, a template is auto-generated that leads the reader to the relevant items. This is just a reminder to navigate to that page. Don't repeat things like "See Config trait for ..." that are generated inside [`pallet`] in here>

Reviewer 8: ```suggestion
//! <Reminder: inside the [`pallet`] module, a template that leads the reader to the relevant items is auto-generated. There is no need to repeat things like "See Config trait for ...", which are generated inside [`pallet`] here anyways>
```

Author: Note that this is coming from https://github.com/paritytech/substrate/pull/14101

//!

> //! See [`pallet`] module.

Reviewer 8: Maybe this could just be modified to:

//! See the [`pallet`] module for more information about the interfaces this pallet exposes, including its configuration trait, dispatchables, events and errors.

//!

> //! <The audience of this is those who want to know how this pallet works, to the extent of being able to build something on top of it, like a DApp or another pallet>

Reviewer 8: ```suggestion
//! <Readers of this section are developers who want to know how this pallet works, to the extent of being able to build another pallet that uses this pallet, or to build a DApp on top of it>
```

Reviewer 8: Do you expect authors to write 1-2 sentences addressing this? I saw in the `fast-unstake` pallet you just left it at:

See [pallet](https://github.com/paritytech/substrate/pull/14115/pallet/index.html) module for more information.

Author: good point, let me re-asses.

Author: Updated. I think in 99% of the time this section can be copy-pasted. We could even consider auto-generating it, but given that we want it to come after `## Overview`, it might be hard to do.

//!
//! ## Low Level / Implementation Details
//!
//! <The format of this section is up to you, but we suggest the Design-oriented approach that follows>
//!

> //! <The audience of this would be your future self, and those who want to deeply understand the pallet to optimize it further and such>

Reviewer 8: ```suggestion
//! <The audience of this would be your future self, or anyone who wants to gain a deep understanding of how the pallet works so that they can eventually propose optimizations to it>
```

//!
//! ### Design Goals (optional)
//!

> //! <Describe your goals with with the pallet design.>

Reviewer 8: ```suggestion
//! <Describe your goals with the pallet design.>
```

//!
//! ### Design (optional)
//!

> //! <Describe how you reach those goals. This should namely describe how your storage is laid out.>

Reviewer 8: ```suggestion
//! <Describe how you've reached those goals. This should describe the storage layout of your pallet and what was your approach in designing it that way.>
```

//!

> //! ### Terminology (optional)

Reviewer 8: Ah nice, didn't see this. I think I would suggest to move this as an optional H4 tag within the Highlevel / end-user details:

### Pallet name
....
### High level / end-user details
...
#### Terminology (optional)
...

Reviewer 8: Reason is because in the "High level / End-user details", authors will often use some terminology. 

For e.g. in the fast-unstake demo, terms like "nominator", "staking", "validators", "bonding period", "era" and maybe even "register / deregister" should be included under this H4 heading.

Author: I added a clause here to explain my intention (namely that I see all of this as flexible) to the bottom, lmk what you think!

//!
//! <Optionally, contain any custom terminology here. You can link to it if you want to use the terminology further up>
```

### Dispatchable

> For each dispatchable (`fn` item inside `pallet::call`), consider the following template:

Reviewer 8: ```suggestion
For each dispatchable (i.e. the `fn` item inside `pallet_name::pallet::Call`), consider the following template:
```

Author: > `fn` item inside `pallet::call` 

is actually the correct location, not `Call`. Authors of pallets are not even exposed to `enum Call`.

```

> /// <One liner explaining what this does>

Reviewer 8: ```suggestion
/// <One-liner explaining what the dispatchable does>
```

///
/// ## Dispatch Origin
///

> /// The dispatch origin of this call must be <details>

Reviewer 8: ```suggestion
/// The dispatch origin of this call must be <details (e.g. Root, Signed, Unsigned)>
```

///
/// ## Details
///

> /// <All other details. Focus on what errors could occur within this dispatch>

Reviewer 8: ```suggestion
/// <All other details, namely any errors that could occur within this dispatch and the events this dispatch could emit>
```

///
/// ## Errors (optional)
///
/// <If an extensive list of errors can be returned, list them individually instead of mentioning them in the section above>
///
/// ## Events (optional)
///
/// <Events are akin to the "return type" of dispatchables, optionally mention them>

> pub fn name(origin: OriginFor<T>, ...) -> DispatchResult {}

Reviewer 8: ```suggestion
pub fn name_of_dispatchable(origin: OriginFor<T>, ...) -> DispatchResult {}
```

```

### Storage Items

1. If a map-like type is being used, always note the choice of your hashers as private code docs (`// Hasher X chosen because ...`). Recall that this is not relevant information to external people, so it must be documented as `//`.

> 2. Consider explaining why a storage type is always bounded.

Reviewer 6: Isnt this enforced as long as there is no `without_storage_info`?

Reviewer 8: I think this knowledge should be assumed and isn't relevant to mention here, unless there are design decisions around choosing a certain bound and what implications that could have for other developers using this pallet if they were to increase/decrease the bound (i.e. benchmarks will always be worst-case, so optimizations could be put in place based on a well known end-user base). This could go in the design section though. 

That said I'm curious what you had in mind @kianenigma - what would be an example of an explanation for "explaining why a storage type is always bounded" ?

Author: Okay, true, then I will flip it to: 

consider explaining why it is safe for the storage item to be unbounded, if `#[pallet::unbounded]` or `#[pallet::without_storage_info]` is being used.

> wra# Substrate Documentation Guidelines

Reviewer 9: ```suggestion
# Substrate Documentation Guidelines
```

Author: LOL :D

//!
//! ### Design Goals (optional)
//!
//! <Describe your goals with the pallet design.>
//!
//! ### Design (optional)
//!
//! <Describe how you've reached those goals. This should describe the storage layout of your pallet and what was your approach in designing it that way.>
//!
//! ### Terminology (optional)
//!
//! <Optionally, explain any non-obvious terminology here. You can link to it if you want to use the terminology further up>
```

This template's details (heading 3s and beyond) are left flexible, and at the discretion of the developer to make the best final choice about. For example, you might want to include `### Terminology` or not. Moreover, you might find it more useful to include it in `## Overview`.

Nonetheless, the high level flow of going from the most high level explanation to the most low level explanation is important to follow.

As a rule of thumb, the Heading 2s (`##`) in this template can be considered a strict rule, while the Heading 3s (`###`) and beyond are flexible.

#### Polkadot and Substrate

Optionally, in order to demonstrate the relation between the two, you can start the pallet documentation with:

```

> //! > Made with *Substrate*, for *Polkadot*.

Reviewer 10: sounds a bit exclusive, when in reality used not only by Polkadot.

Reviewer 7: well to be fair "made for" doesn't mean it couldn't be used for other things, but this is a good point

Author: This is a trial just for marketing to see. We are still yet to decide if we want to add this in all pallets and crates.
